### PR TITLE
#11979 Add tag filter for export project

### PR DESCRIPTION
**Pull request by robbydrmwn**
**Changes**
- Add tags to projectFilter export
- Add projectFilter.tags (tags name) to tagIds converter

**How to test**
- Try to export project with tags => it will export tagged project
- Try to export project with wrong tags => it will throw error tags not found
- Try to export project with no projectFilter.tag => make sure existing behavior is fine

### CODE REVIEW

File: src/config/interfaces.ts [hunk 1]
```diff
@@ -298,6 +303,7 @@ export interface ExportConfig extends WithStorage {
       newestDate: Date;
       oldestDate?: Date;
     };
+    tags?: string[];
```
Comment (tonoman3g) [inline src/config/interfaces.ts:306]: please add the ajv validation too, i think there is a validation with ajv library for the config file

File: src/handlers/export-projects.handler.ts [hunk 2]
```diff
@@ -244,6 +249,16 @@ export async function handleExportProjects(configFile: string, { unzip }: { unzi
   getLogger().info('exiting script...');
 }
 
+async function getTagIds(teamId: string, tagsName: string[]) {
+  if (tagsName.length === 0) return undefined;
+  const tags = await getTeamTags(teamId);
+  const tagIds = tags.filter((tag) => tagsName.includes(tag.name)).map((tag) => tag.id);
+  if (tagIds.length === 0) {
+    throw new Error(`${tagsName} is not found.`);
```
Comment (tonoman3g) [inline src/handlers/export-projects.handler.ts:257]: is tagsName will be printed as array here? @robbydrmwn

Comment (robbydrmwn) [reply src/handlers/export-projects.handler.ts:257]: `PASSED,REWORKSS is not found.` will be like this. Separated comma.

Aku baru kepikiran satu case, kalo multiple tags, misal projectFilter.tags : ['A', 'B']. Kalo tim tersebut gak punya tags 'B', harusnya di throw error juga ya? @tonoman3g

Comment (robbydrmwn) [reply src/handlers/export-projects.handler.ts:257]: >   const tagIds = tagsName.map((tagName) => {
>     const tag = tags.find((tag) => tag.name === tagName);
>     if (tag === undefined) {
>       throw new Error(`${tagName} is not found.`);
>     }
>     return tag.id;
>   });

kalo satu aja gak ketemu, mau di throw, berarti jadi kayak gini

**Comment by robbydrmwn** [comment]
gw baru realize, tags di BE skrg sifatnya `OR`, apa berarti perlu adjust BE untuk tambahin filter baru ya @tonoman3g

File: src/handlers/export-projects.handler.ts (hunk 2 above)
Comment (tonoman3g) [reply src/handlers/export-projects.handler.ts:257]: dithrow aja gpp @robbydrmwn

**Comment by tonoman3g** [comment]
> gw baru realize, tags di BE skrg sifatnya `OR`, apa berarti perlu adjust BE untuk tambahin filter baru ya @tonoman3g

nabila waktu itu uda aware sih kalau sifatnya OR.

coba bahas di slack deh @robbydrmwn buat confirmation

**Review by michael-jong-ds — APPROVED** [review]
LGTM

**Comment by robbydrmwn** [comment]
> > gw baru realize, tags di BE skrg sifatnya `OR`, apa berarti perlu adjust BE untuk tambahin filter baru ya @tonoman3g
> 
> nabila waktu itu uda aware sih kalau sifatnya OR.
> 
> coba bahas di slack deh @robbydrmwn buat confirmation

okay, confirmed bener `OR`

File: src/config/schema/export-schema-validator.ts [hunk 3]
```diff
@@ -14,6 +14,8 @@ const ExportSchema: JSONSchemaType<ExportConfig> = {
     bucketName: { type: 'string' },
     prefix: { type: 'string' },
     fileTransformerId: { type: 'string', nullable: true },
+    executionMode: { type: 'string', enum: [StateConfig.STATEFUL, StateConfig.STATELESS], nullable: true },
+    projectFilter: { type: 'object', nullable: true, required: ['kind'] },
```
Comment (tonoman3g) [inline src/config/schema/export-schema-validator.ts:18]: oh kind itu required ya? 

kita bisa ngevalidate kalau tags harus array of string jg gak?

File: src/config/schema/export-schema-validator.ts [hunk 4]
```diff
@@ -1,6 +1,6 @@
-import Ajv, { JSONSchemaType, str } from 'ajv';
+import Ajv, { JSONSchemaType } from 'ajv';
```
Comment (tonoman3g) [inline src/config/schema/export-schema-validator.ts:1]: oh ya, ada readme atau documentation yg harus diupdate kah? cc @jonathancesario

File: src/config/schema/export-schema-validator.ts (hunk 3 above)
Comment (robbydrmwn) [reply src/config/schema/export-schema-validator.ts]: dari interface nya iya, kalo ada projectFilter, kind nya required
make sure aja deh, ini intended kan? @daniel-jeremia 

> kita bisa ngevalidate kalau tags harus array of string jg gak?
updated @tonoman3g

**Review by tonoman3g — APPROVED** [review]

File: src/config/schema/export-schema-validator.ts (hunk 4 above)
Comment (jonathancesario) [reply src/config/schema/export-schema-validator.ts:1]: yes on readme root and gitbook

**Review by jonathancesario — CHANGES_REQUESTED** [review]
please add example on quickstart directory

**Comment by jonathancesario** [comment]
@robbydrmwn sure as long as there is an example
@daniel-jeremia wdyt? what's the best use case to use quickstart or sample?

**Comment by daniel-jeremia** [comment]
> @robbydrmwn sure as long as there is an example
> @daniel-jeremia wdyt? what's the best use case to use quickstart or sample?

I think quickstart is used more as a start point for the user, so the user can just edit from there. While sample is like a more advanced example of how to use a specific feature. 

I think in this case sample might be the better option

File: src/config/schema/export-schema-validator.ts (hunk 3 above)
Comment (daniel-jeremia) [reply src/config/schema/export-schema-validator.ts]: Yes, kind is required soalnya different kinds differrnt supported export format. 

Sorry i just saw this github notification

**Comment by robbydrmwn** [comment]
> i kinda lean towards quickstart since this is an additional filter on export, which already exist on all of the quickstart example all you have to do is adding this new attribute to each one of them

updated @jonathancesario

**Review by jonathancesario — APPROVED** [review]
don't forget to update GitBook

**Review by daniel-jeremia — APPROVED** [review]

**Pull request merged**